### PR TITLE
Add React router tests

### PR DESCRIPTION
**Pull request by gabalafou**
This PR contains the tests for https://github.com/conda-incubator/conda-store-ui/pull/429.

It is a separate PR because it makes significant changes to the React app code.

I believe that these changes are not only necessary to enable these tests but that they also fix a long-standing bug that I guess we did not even realize that we had. 

## Description

This pull request:

- Swaps the order in which config variables are read. First it reads from `condaStoreConfig`, then from `process.env` (environment variables). Important note: these environment variables can only be read at build time.
- Adds a new Playwright test file for router testing
- Adds an HTML file that allows Playwright to load the app with memory routing configured
- Configures Webpack to output the HTML file

## Pull request checklist

- [x] Did you test this change locally?
- [ ] Did you update the documentation (if required)?
- [x] Did you add/update relevant tests for this change (if required)?

### CODE REVIEW

File: src/preferences.tsx [hunk 1]
```diff
@@ -17,50 +17,50 @@ export interface IPreferences {
...
-  typeof window !== "undefined" && (window as any);
+let condaStoreConfig: any = {};
+if (typeof window !== "undefined" && "condaStoreConfig" in window) {
+  condaStoreConfig = window.condaStoreConfig;
+}
 
 export const prefDefault: Readonly<IPreferences> = {
   apiUrl:
-    process.env.REACT_APP_API_URL ??
     condaStoreConfig.REACT_APP_API_URL ??
```
Comment (gabalafou) [inline src/preferences.tsx:31]: The Conda Store docs say that [condaStoreConfig can be used to configure the UI app at runtime](https://conda.store/conda-store-ui/how-tos/configure-ui). 

I take this to also imply that run-time should override build-time configuration. 

If not, it is impossible to set a config variable in the way that I do in the `memory-router-test.html` file in this PR if that variable was already defined as an environment variable at build time.

So for example, if in my .env file I have:

```
REACT_APP_ROUTER_TYPE=browser
```

And then in my app index html page I have:

```
<script>
window.condaStoreConfig = {
  REACT_APP_ROUTER_TYPE: "memory"
};
<script>
```

Then the app will actually be loaded with the browser router and not the memory router. I don't think that's what we want, but I may be mistaken.

Comment (trallard) [reply src/preferences.tsx:31]: Runtime indeed overwrites/supersedes build time.
This is now how we are passing default config values to conda-store-server vs the hack we used to have before https://github.com/conda-incubator/conda-store/blob/d39e5b47577aa6764b1bf99a0ab675fe8782e969/conda-store-server/conda_store_server/_internal/server/templates/conda-store-ui.html#L11

File: test/playwright/test_react_router.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    expect(page.get_by_text("404")).to_be_visible()
+
+
+def test_memory_router_404_not_found(page: Page, test_config):
+    """The memory router has been configured to load the root view at any route
+    """
+    # The route `/memory-router-test.html` is not a route recognized by the
+    # React app. With the browser router, an unknown route would give a 404.
+    page.goto(test_config["base_url"] + "/memory-router-test.html")
+    expect(page.get_by_test_id("no-environment-selected")).to_be_visible()
```
Comment (trallard) [inline test/playwright/test_react_router.py:50]: So basically, do non-existing routes always redirect to `/conda-store`?

Comment (gabalafou) [reply test/playwright/test_react_router.py:50]: Not exactly. The memory router does not redirect anywhere. It loads the app at the route you pre-configure. For example, in my other PR, I configure the [memory router to start with the root route](https://github.com/conda-incubator/conda-store-ui/pull/429/files#diff-26ad4b834941d9b19ebf9db8082bd202aaf72ea0ddea85f5a8a0cb3c729cc6f2R73). That means that no matter what the URL in the browser address bar says, if the app is loaded at that URL, it will load the root view.

The browser router, by contrast, tries to read the URL path from the browser address bar when the app loads and then render the app view that matches that route (or render a 404 if no view matches the route).

In other words, the browser router is URL-aware, whereas the memory router is not. That's the main difference.

File: src/preferences.tsx (hunk 1 above)
Comment (gabalafou) [reply src/preferences.tsx:31]: That's interesting. Does that file you sent actually work? Because as far as I can tell, the only way it would work is if the `main.js` file at `static/conda-store-ui/main.js` were built in an environment in which none of those `REACT_APP_`-prefixed variables were defined at build time. I searched the UI codebase for `condaStoreConfig` to see if I was missing something, but the only place that variable is read is in the `preferences.tsx` file, and in that file, it's pretty clear that the variable from `condaStoreConfig` is only read if the corresponding environment variable were nullish at build-time.

For example, let's look at the following JavaScript:

```js
process.env.REACT_APP_API_URL ??
condaStoreConfig.REACT_APP_API_URL ??
"http://localhost:8080/conda-store/"
```

At build time Webpack will do a string substitution and replace `process.env.REACT_APP_API_URL` with the value in the environment, if it's defined. So, if it's set to "http://localhost:8080/conda-store/", then the above JavaScript is transformed to:

```js
"http://localhost:8080/conda-store/" ??
condaStoreConfig.REACT_APP_API_URL ??
"http://localhost:8080/conda-store/"
```

Since the first argument in the above expression is not "nullish", meaning neither `null` nor `undefined`, then the overall expression reduces to that first argument, i.e.:

```js
"http://localhost:8080/conda-store/"
```

On the other hand, if that environment variable is not defined at build time, the source code is transformed as:

```js
undefined ??
condaStoreConfig.REACT_APP_API_URL ??
"http://localhost:8080/conda-store/"
```

Since the left hand side of the nullish operator is undefined the runtime will try to read the value from `condaStoreConfig`, and only if that too is null or undefined, will it return the last operand.

So the only way the condaStoreConfig variable plays any role at run time is if the corresponding environment variable was not defined at build time.
